File: apps/cli/src/tool-migration-processes.ts
import { execSync } from 'node:child_process';
import {
  getToolMigrationProfile,
  type MigrateToolId,
} from './tool-migration-profiles.js';

const TOOL_PROCESS_IMAGE_NAMES: Partial<Record<MigrateToolId, readonly string[]>> = {
  cursor: ['Cursor.exe'],
  'cursor-roaming': ['Cursor.exe'],
  'cursor-local': ['Cursor.exe'],
  vscode: ['Code.exe'],
  'claude-code': ['claude.exe'],
  'codex-cli': ['codex.exe'],
  'claude-desktop': ['Claude.exe'],
  'docker-desktop': ['Docker Desktop.exe', 'com.docker.backend.exe', 'com.docker.build.exe'],
};

/** Windows image names to check before migration (deduped). */
export function processImageNamesForTool(id: MigrateToolId): readonly string[] {
  const profile = getToolMigrationProfile(id);
  if (profile.bundleLegs?.length) {
    const names = new Set<string>();
    for (const leg of profile.bundleLegs) {
      for (const name of processImageNamesForTool(leg.sourceProfileId as MigrateToolId)) {
        names.add(name);
      }
    }
    return [...names];
  }
  return TOOL_PROCESS_IMAGE_NAMES[id] ?? [];
}

function isWindowsProcessRunning(imageName: string): boolean {
  try {
    const out = execSync(`tasklist /FI "IMAGENAME eq ${imageName}" /NH`, {
      encoding: 'utf8',
      timeout: 15_000,
    });
    if (/INFO:\s*No tasks are running/i.test(out)) return false;
    return out.toLowerCase().includes(imageName.toLowerCase());
  } catch {
    return false;
  }
}

/** Image names currently running for this tool profile (Windows only). */
export function detectRunningToolProcesses(id: MigrateToolId): string[] {
  if (process.platform !== 'win32') return [];
  return processImageNamesForTool(id).filter((name) => isWindowsProcessRunning(name));
}

export function runningProcessWarning(running: readonly string[]): string | null {
  if (running.length === 0) return null;
  return `Close these processes before Run migration: ${running.join(', ')} (check Task Manager and the tray icon).`;
}

export function enrichPlanWithRunningProcesses<
  T extends { readonly warnings: readonly string[]; readonly running_processes?: readonly string[] },
>(
  tool: MigrateToolId | undefined,
  plan: T,
): T {
  if (!tool || process.platform !== 'win32') return plan;
  const running = detectRunningToolProcesses(tool);
  const warning = runningProcessWarning(running);
  if (!warning) return { ...plan, running_processes: [] };
  return {
    ...plan,
    running_processes: [...running],
    warnings: plan.warnings.includes(warning) ? plan.warnings : [...plan.warnings, warning],
  };
}

File: apps/desktop/src-tauri/src/util/tool_migration_processes.rs
use super::tool_migration::ToolId;

#[cfg(windows)]
fn is_process_running(image_name: &str) -> bool {
    use std::process::Command;
    let output = Command::new("tasklist")
        .args(["/FI", &format!("IMAGENAME eq {image_name}"), "/NH"])
        .output();
    let Ok(output) = output else {
        return false;
    };
    let text = String::from_utf8_lossy(&output.stdout);
    if text.contains("INFO: No tasks are running") {
        return false;
    }
    text.to_ascii_lowercase().contains(&image_name.to_ascii_lowercase())
}

#[cfg(not(windows))]
fn is_process_running(_image_name: &str) -> bool {
    false
}

fn process_image_names(tool: ToolId) -> Vec<&'static str> {
    match tool {
        ToolId::Cursor | ToolId::CursorRoaming | ToolId::CursorLocal => vec!["Cursor.exe"],
        ToolId::Vscode => vec!["Code.exe"],
        ToolId::ClaudeCode => vec!["claude.exe"],
        ToolId::CodexCli => vec!["codex.exe"],
        ToolId::ClaudeDesktop => vec!["Claude.exe"],
        ToolId::GoogleChrome => vec!["chrome.exe"],
        ToolId::MicrosoftEdge => vec!["msedge.exe"],
        ToolId::Brave => vec!["brave.exe"],
        ToolId::Firefox => vec!["firefox.exe"],
        ToolId::Discord | ToolId::DiscordRoaming | ToolId::DiscordLocal => vec!["Discord.exe"],
        ToolId::Spotify => vec!["Spotify.exe"],
        ToolId::Slack => vec!["slack.exe"],
        ToolId::Telegram => vec!["Telegram.exe"],
        ToolId::Notion => vec!["Notion.exe"],
        ToolId::ObsStudio => vec!["obs64.exe", "obs32.exe"],
        ToolId::EpicGames => vec!["EpicGamesLauncher.exe"],
        ToolId::SteamAppdata => vec!["steam.exe"],
        ToolId::BattleNet => vec!["Battle.net.exe"],
        ToolId::DockerDesktop => vec![
            "Docker Desktop.exe",
            "com.docker.backend.exe",
            "com.docker.build.exe",
        ],
        ToolId::NpmCache | ToolId::PnpmStore => vec![],
    }
}

pub fn detect_running_processes(tool: ToolId) -> Vec<String> {
    process_image_names(tool)
        .into_iter()
        .filter(|name| is_process_running(name))
        .map(|s| s.to_string())
        .collect()
}

pub fn running_process_warning(tool: ToolId) -> Option<String> {
    let running = detect_running_processes(tool);
    if running.is_empty() {
        return None;
    }
    Some(format!(
        "Close these processes before Run migration: {} (check Task Manager and the tray icon).",
        running.join(", ")
    ))
}
